EIPS/eip-6327.md

```diff
@@ -0,0 +1,316 @@
+- ES can also be used as a plugin to a smart contract wallet, like Account Abstraction [ERC-4337](./eip-4337.md). A decentralized password is picked instead of the private key. This could lead to a smooth onboarding experiences for new Ethereum Dapp users.
+
+
+## Specification
+
+Let:
+
+- `pwdhash` represents the hash of the private secret (password).
+- `datahash` represents the hash of an intended transaction data.
+- `fullhash` represents the hash of `datahash` and all the well-known variables.
+- `expiration` is the timestamp after which the intended transaction expires. 
+- `allhash` represents the hash of `fullhash` and `pwdhash`.
+
+
+There are three parties involved, Verifier, Requester and Prover.
+
+- A verifier, 
+  - SHOULD compute `fullhash` from a `datahash`, which is provided by the requester.
+  - SHOULD derive `pwdhash` for a given address. The address can be an EOA or a smart contract wallet.
+  - SHOULD verify the proof with the derived `pwdhash`, the computed `fullhash` and a `allhash`, which is submitted by the requester.
+- A requester
+  - SHOULD generate `datahash` and decide an `expiration`.
+  - SHALL request a verification from the verifier with, 
+    - `proof` and `allhash` which are provided by the prover;
+    - `datahash`;
+    - `expiration`.
+- A prover
+  - SHOULD generate the `proof` and `allhash` from, 
+    - `datahash` and `expiration` which are agreed with the requester;
+    - `nonce` and other well-known variables. 
+
+There are also some requirements.
+
+- well-known variable SHOULD be available to all parties.
+  - SHOULD include a `nonce`.
+  - SHOULD include a `chainid`.
+  - MAY include any variable that is specific to the verifier.
+- public statements SHOULD include, 
+  - one reflecting the `pwdhash`;
+  - one reflecting the `fullhash`;
+  - one reflecting the `allhash`.
+- The computation of `fullhash` SHOULD be agreed by both the verifier and the prover.
+- The computation of `datahash`
+
+### `IElasticSignature` Interface 
+
+This is the verifier interface.
+
+```solidity
+pragma solidity ^0.8.0;
+
+interface IElasticSignature {
+    /**
+     * Event emitted after user set/reset their password
+     * @param user - an user's address, for whom the password hash is set. It could be a smart contract wallet address
+     *  or an EOA wallet address.
+     * @param pwdhash - a password hash
+     */
+    event SetPassword(address indexed user, uint indexed pwdhash);
+
+    /**
+     * Event emitted after a successful verification performed for an user
+     * @param user - an user's address, for whom the submitted `proof` is verified. It could be a smart contract wallet
+     *  address or an EOA wallet address.
+     * @param nonce - a new nonce, which is newly generated to replace the last used nonce. 
```

Comment:
Does it indicate that `nonce` could be non-sequential?

Reply:
No, it can't be non-sequential, because the contract using nonce+1 for the next nonce

Reply:
In that case, help me understand why the event needs to emit nonce, if nonce can be inferred?